# Data Exploration

In [206]:
import pandas as pd
import numpy as np
import os 
import json

In [207]:
# Load the data paths
data_dir = "../data/"

# Load the raw data into a pandas dataframe

test = pd.read_csv(os.path.join(data_dir, "test.csv"))
train = pd.read_csv(os.path.join(data_dir, "train.csv"))
sample = pd.read_csv(os.path.join(data_dir, "sample_submission.csv"))

In [208]:
# Shapes (rows, columns)
print("Train shape:", train.shape)
print("Test shape:", test.shape)
print("Sample submission shape:", sample.shape)

# Column names
print("\nTrain columns:", train.columns.tolist())
print("\nTest columns:", test.columns.tolist())

# First 5 rows (peek at data)
display(train.head())
display(test.head())
display(sample.head())

# Info (dtypes, non-null counts)
print("\nTrain info:")
print(train.info())
print("\nTest info:")
print(test.info())

# Quick descriptive stats for numerics
train.describe().T


Train shape: (75973, 14)
Test shape: (32567, 13)
Sample submission shape: (32567, 2)

Train columns: ['carID', 'Brand', 'model', 'year', 'price', 'transmission', 'mileage', 'fuelType', 'tax', 'mpg', 'engineSize', 'paintQuality%', 'previousOwners', 'hasDamage']

Test columns: ['carID', 'Brand', 'model', 'year', 'transmission', 'mileage', 'fuelType', 'tax', 'mpg', 'engineSize', 'paintQuality%', 'previousOwners', 'hasDamage']


,carID,Brand,model,year,price,transmission,mileage,fuelType,tax,mpg,engineSize,paintQuality%,previousOwners,hasDamage
0,69512,VW,Golf,2016.0,22290,Semi-Auto,28421.0,Petrol,NaN,11.417268,2.0,63.0,4.000000,0.0
1,53000,Toyota,Yaris,2019.0,13790,Manual,4589.0,Petrol,145.0,47.900000,1.5,50.0,1.000000,0.0
2,6366,Audi,Q2,2019.0,24990,Semi-Auto,3624.0,Petrol,145.0,40.900000,1.5,56.0,4.000000,0.0
3,29021,Ford,FIESTA,2018.0,12500,anual,9102.0,Petrol,145.0,65.700000,1.0,50.0,-2.340306,0.0
4,10062,BMW,2 Series,2019.0,22995,Manual,1000.0,Petrol,145.0,42.800000,1.5,97.0,3.000000,0.0


,carID,Brand,model,year,transmission,mileage,fuelType,tax,mpg,engineSize,paintQuality%,previousOwners,hasDamage
0,89856,Hyundai,I30,2022.878006,Automatic,30700.000000,petrol,205.0,41.5,1.6,61.0,3.0,0.0
1,106581,VW,Tiguan,2017.000000,Semi-Auto,-48190.655673,Petrol,150.0,38.2,2.0,60.0,2.0,0.0
2,80886,BMW,2 Series,2016.000000,Automatic,36792.000000,Petrol,125.0,51.4,1.5,94.0,2.0,0.0
3,100174,Opel,Grandland X,2019.000000,Manual,5533.000000,Petrol,145.0,44.1,1.2,77.0,1.0,0.0
4,81376,BMW,1 Series,2019.000000,Semi-Auto,9058.000000,Diesel,150.0,51.4,2.0,45.0,4.0,0.0


,carID,price
0,89856,851000
1,106581,514000
2,80886,323000
3,100174,921000
4,81376,620000



Train info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 75973 entries, 0 to 75972
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   carID           75973 non-null  int64  
 1   Brand           74452 non-null  object 
 2   model           74456 non-null  object 
 3   year            74482 non-null  float64
 4   price           75973 non-null  int64  
 5   transmission    74451 non-null  object 
 6   mileage         74510 non-null  float64
 7   fuelType        74462 non-null  object 
 8   tax             68069 non-null  float64
 9   mpg             68047 non-null  float64
 10  engineSize      74457 non-null  float64
 11  paintQuality%   74449 non-null  float64
 12  previousOwners  74423 non-null  float64
 13  hasDamage       74425 non-null  float64
dtypes: float64(8), int64(2), object(4)
memory usage: 8.1+ MB
None

Test info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32567 entries, 0 to 32566
Dat

,count,mean,std,min,25%,50%,75%,max
carID,75973.0,37986.000000,21931.660338,0.000000,18993.00,37986.0,56979.0,75972.000000
year,74482.0,2017.096611,2.208704,1970.000000,2016.00,2017.0,2019.0,2024.121759
price,75973.0,16881.889553,9736.926322,450.000000,10200.00,14699.0,20950.0,159999.000000
mileage,74510.0,23004.184088,22129.788366,-58540.574478,7423.25,17300.0,32427.5,323000.000000
tax,68069.0,120.329078,65.521176,-91.121630,125.00,145.0,145.0,580.000000
mpg,68047.0,55.152666,16.497837,-43.421768,46.30,54.3,62.8,470.800000
engineSize,74457.0,1.660136,0.573462,-0.103493,1.20,1.6,2.0,6.600000
paintQuality%,74449.0,64.590667,21.021065,1.638913,47.00,65.0,82.0,125.594308
previousOwners,74423.0,1.994580,1.472981,-2.345650,1.00,2.0,3.0,6.258371
hasDamage,74425.0,0.000000,0.000000,0.000000,0.00,0.0,0.0,0.000000


In [209]:
# Description of train
train.describe().T

,count,mean,std,min,25%,50%,75%,max
carID,75973.0,37986.000000,21931.660338,0.000000,18993.00,37986.0,56979.0,75972.000000
year,74482.0,2017.096611,2.208704,1970.000000,2016.00,2017.0,2019.0,2024.121759
price,75973.0,16881.889553,9736.926322,450.000000,10200.00,14699.0,20950.0,159999.000000
mileage,74510.0,23004.184088,22129.788366,-58540.574478,7423.25,17300.0,32427.5,323000.000000
tax,68069.0,120.329078,65.521176,-91.121630,125.00,145.0,145.0,580.000000
mpg,68047.0,55.152666,16.497837,-43.421768,46.30,54.3,62.8,470.800000
engineSize,74457.0,1.660136,0.573462,-0.103493,1.20,1.6,2.0,6.600000
paintQuality%,74449.0,64.590667,21.021065,1.638913,47.00,65.0,82.0,125.594308
previousOwners,74423.0,1.994580,1.472981,-2.345650,1.00,2.0,3.0,6.258371
hasDamage,74425.0,0.000000,0.000000,0.000000,0.00,0.0,0.0,0.000000


In [210]:
# Split the training data into features and target variable (price)
x_train = train.drop(columns=['price'])
y_train = train['price']

In [211]:
x_test = test.copy()

### We check for missing values and wrong values in the column Brand

In [212]:
# Look at the Missing values in x and y
missing_values_x = X.isnull().sum()
missing_values_y = y.isnull().sum()

print("\nMissing values in X:\n", missing_values_x[missing_values_x > 0])
print("\nMissing values in y:\n", missing_values_y[missing_values_y > 0])


Missing values in X:
 model             6715
year              1486
transmission      2687
mileage           1457
fuelType          3012
tax               7898
mpg               7923
engineSize        1515
paintQuality%     1520
previousOwners    1547
hasDamage         1544
dtype: int64

Missing values in y:
 []


- We start with fixing values we can do without predicting them with statistical or other methods. We start with mapping with JSON files. And we will delete all the values which make no sense 

In [213]:
# unique brands in the column brand
unique_brands = X['Brand'].unique()
print("\nUnique brands in 'Brand':\n", unique_brands)


Unique brands in 'Brand':
 ['Volkswagen' 'Toyota' 'Audi' 'Ford' 'BMW' 'Škoda' 'Opel' 'Mercedes-Benz'
 'Hyundai' 'Skoda' 'Mercedes']


### First we fix the brand names and afterwards we will map the brands from the model names

In [214]:
# Load the JSON mapping
with open("../mapping/brandname_mapping.json", "r", encoding="utf-8") as f:
    CANON = json.load(f)

In [215]:
import re

# Normalization function
def norm_brand(brand):
    if pd.isna(brand):
        return brand
    brand = brand.strip().lower()
    brand = re.sub(r'[.,-_]', ' ', brand)
    brand = ' '.join(brand.split())  # Remove extra spaces
    return brand

In [216]:
# Cleaning function using imported JSON
def clean_brand(col: pd.Series) -> pd.Series:
    nb = col.apply(norm_brand)
    mapped = nb.map(CANON)
    return mapped

In [217]:
# apply normalization and canonical mapping to the Brand column in x_train and x_test
x_train['Brand'] = x_train['Brand'].apply(norm_brand).map(CANON)
x_test['Brand'] = x_test['Brand'].apply(norm_brand).map(CANON)

In [218]:
# unique brands in the column brand
unique_brands = x_train['Brand'].unique()
print("\nUnique brands in 'Brand':\n", unique_brands)


Unique brands in 'Brand':
 ['Volkswagen' 'Toyota' 'Audi' 'Ford' 'BMW' 'Škoda' 'Opel' 'Mercedes-Benz'
 'Hyundai' nan]


In [219]:
# Show the entries with missing values in brand
missing_brand_entries = x_train[x_train['Brand'].isnull()]
display(missing_brand_entries)

,carID,Brand,model,year,transmission,mileage,fuelType,tax,mpg,engineSize,paintQuality%,previousOwners,hasDamage
56,66591,NaN,T-Roc,2018.0,Semi-Auto,584.0,Petrol,150.0,41.5,2.0,79.0,1.0,0.0
75,5034,NaN,A3,2016.0,Manual,52519.0,Diesel,20.0,72.4,1.6,60.0,2.0,0.0
81,33153,NaN,I20,2020.0,MANUAL,3500.0,Petrol,145.0,47.9,1.0,41.0,2.0,0.0
176,45338,NaN,CL Class,2020.0,Semi-Auto,NaN,Petrol,145.0,33.2,2.0,61.0,4.0,0.0
191,66712,NaN,POLO,2016.0,Manual,32800.0,Petrol,20.0,60.1,1.0,54.0,4.0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...
75632,72615,NaN,Golf,2015.0,Manual,79000.0,Diesel,20.0,67.3,2.0,68.0,0.0,0.0
75720,3436,NaN,Q5,2015.0,Semi-Auto,18188.0,Diesel,200.0,47.1,2.0,84.0,3.0,0.0
75810,7392,NaN,A5,2019.0,Automatic,10.0,Petrol,145.0,39.8,2.0,70.0,1.0,0.0
75828,13545,NaN,5 Series,2019.0,Semi-Auto,7609.0,Diese,145.0,65.7,2.0,67.0,3.0,0.0


- We fixed wrong names in the column Brand but we have missing values. We will try to fill them by looking at the column model.
But there are also some wrong values in the column model. We will try to fix them first


In [220]:
# Count the unique values of model in x
unique_models = x_train['model'].value_counts()
print("\nCounts of unique values in 'model' for entries with missing 'Brand':\n", unique_models)

# save the unique_models to a json file
unique_models.to_json("../mapping/unique_model_counts.json")


Counts of unique values in 'model' for entries with missing 'Brand':
 model
 Focus        6353
 C Class      4886
 Fiesta       4130
 Golf         3081
 Corsa        2182
              ... 
UP               1
 scal            1
 Fox             1
astra            1
 glb class       1
Name: count, Length: 735, dtype: int64


- Canonical Mapping and Normalization for model names

In [221]:
import re
import pandas as pd
import numpy as np
import json

def norm_model(x: object) -> str | float:
    """Lowercase, keep [a-z0-9+], strip spaces/hyphens/underscores."""
    if pd.isna(x):
        return np.nan
    s = str(x).lower()
    s = re.sub(r'[\s\-_]+', '', s)           # drop spaces, hyphens, underscores
    s = re.sub(r'[^a-z0-9\+]', '', s)        # keep letters, digits, plus
    return s or np.nan


In [222]:
# Load mapping JSON
with open("../mapping/modelname_mapping.json", "r", encoding="utf-8") as f:
    MM = json.load(f)

ALIASES = MM.get("aliases", {})
REGEX_RULES = MM.get("regex_rules", [])

In [223]:
def apply_regex_rules(norm_key: str) -> str:
    for rule in REGEX_RULES:
        pat = rule["pattern"]
        rep = rule["replace"]
        norm_key = re.sub(pat, rep, norm_key)
    return norm_key

In [224]:
def clean_model(series: pd.Series) -> pd.Series:
    normed = series.apply(norm_model)
    normed = normed.apply(lambda x: apply_regex_rules(x) if isinstance(x, str) else x)
    return normed.map(ALIASES)

In [225]:
# Apply the clean model function to the 'model' column for train and test
x_train['model'] = clean_model(x_train['model'])
x_test['model'] = clean_model(x_test['model'])

# Apply the apply_regex_rules function to the 'model' column
x_train['model'] = x_train['model'].apply(lambda x: apply_regex_rules(x) if isinstance(x, str) else x)
x_test['model'] = x_test['model'].apply(lambda x: apply_regex_rules(x) if isinstance(x, str) else x)

# unique models in the column model for x_train
unique_models = X['model'].unique()
print("\nUnique models in 'model':\n", unique_models)
# print the number of unique models
print("\nNumber of unique models in 'model':", len(unique_models))


Unique models in 'model':
 ['Golf' 'Yaris' 'Q2' 'Fiesta' '2 Series' '3 Series' 'A3' 'Octavia'
 'Passat' 'Focus' 'Insignia' nan 'Q3' 'Fabia' 'A-Class' 'Ka+' 'GLC' 'i30'
 'C-Class' 'E-Class' 'Q5' 'Up' 'Mokka X' 'Corsa' 'Astra' 'TT' '5 Series'
 'Aygo' '4 Series' 'SLK' 'T-Roc' 'EcoSport' 'Tucson' 'X-Class' 'ix20'
 'i20' 'A1' 'Auris' 'Sharan' 'Adam' 'X3' 'A8' 'GLS' 'B-MAX' 'A4' 'Kona'
 'i10' 'Mokka' 'S-MAX' 'X2' 'Crossland X' 'Tiguan' 'A5' 'GLE' 'Zafira'
 'Ioniq' 'A6' 'Mondeo' 'Yeti Outdoor' 'X1' 'Scala' 'S-Class' '1 Series'
 'Kamiq' 'Kuga' 'Tourneo Connect' 'Q7' 'GLA' 'SL' 'Grandland X' 'i800'
 'RAV4' 'Touran' 'Citigo' 'Prius' 'Corolla' 'B-Class' 'Kodiaq' 'V-Class'
 'Caddy Maxi Life' 'Superb' 'Getz' 'Beetle' 'Galaxy' 'M3' 'GTC' 'X4' 'Ka'
 'ix35' 'Grand Tourneo Connect' 'M4' 'Tourneo Custom' 'Z4' 'X5' 'Meriva'
 'RS6' 'Verso' 'CLS' 'C-MAX' 'CLA' 'i40' 'Tiguan Allspace' '6 Series'
 'Caravelle' 'Karoq' 'i3' 'Grand C-MAX' 'T-Cross' 'A7' 'Agila' 'GT86'
 'Yeti' 'California' 'Land Cruiser' 'Edge'

In [226]:
# Count all the entries in the json to compare with the unique models in the column model
with open("../mapping/modelname_mapping.json", "r", encoding="utf-8") as f:
    MM = json.load(f)
ALIASES = MM.get("aliases", {})
print("\nNumber of entries in the modelname_mapping.json:", len(ALIASES))


Number of entries in the modelname_mapping.json: 195


- Now we have all the modelnames fixed and normalized. We can use this to fill the missing values in the column Brand.

In [227]:
# Check for missing values in 'Brand' before filling
print("\nMissing values in 'Brand' before filling:\n", x_train['Brand'].isnull().sum())

# Load mapping
with open("../mapping/brand_model_mapping.json", "r", encoding="utf-8") as f:
    brand_to_model = json.load(f)

# Map directly using the mapping
x_train["Brand"] = x_train["Brand"].fillna(x_train["model"].map(brand_to_model))

# Check for missing values in 'Brand' after filling
print("\nMissing values in 'Brand' after filling:\n", x_train['Brand'].isnull().sum())


Missing values in 'Brand' before filling:
 1539

Missing values in 'Brand' after filling:
 150


In [228]:
# Check for nan_brand values with models that could not be mapped
nan_brand_models = x_train[x_train['Brand'].isnull()]['model'].unique()
print("\nModels with missing 'Brand' after mapping:\n", nan_brand_models)


Models with missing 'Brand' after mapping:
 [nan]


- We have mapped every possible model to its brand. The remaining missing values in the column Brand are due to missing model names.
  

In [229]:
# Look at the Missing values in x and y
missing_values_x = x_train.isnull().sum()
missing_values_y = y_train.isnull().sum()

print("\nMissing values in X:\n", missing_values_x[missing_values_x > 0])
print("\nMissing values in y:\n", missing_values_y[missing_values_y > 0])



Missing values in X:
 Brand              150
model             6865
year              1491
transmission      1522
mileage           1463
fuelType          1511
tax               7904
mpg               7926
engineSize        1516
paintQuality%     1524
previousOwners    1550
hasDamage         1548
dtype: int64

Missing values in y:
 []


In [230]:
# Show the entries with missing values in Brand
missing_brand_entries = x_train[x_train['Brand'].isnull()]
display(missing_brand_entries)

,carID,Brand,model,year,transmission,mileage,fuelType,tax,mpg,engineSize,paintQuality%,previousOwners,hasDamage
176,45338,NaN,NaN,2020.0,Semi-Auto,NaN,Petrol,145.0,33.2,2.0,61.000000,4.0,0.0
191,66712,NaN,NaN,2016.0,Manual,32800.0,Petrol,20.0,60.1,1.0,54.000000,4.0,0.0
241,73298,NaN,NaN,2019.0,Semi-Auto,3797.0,diesel,145.0,44.1,2.0,77.000000,2.0,0.0
442,16342,NaN,NaN,2017.0,MANUAL,28820.0,Diesel,0.0,74.4,1.5,NaN,0.0,0.0
2031,12064,NaN,NaN,2018.0,Automatic,11258.0,NaN,145.0,64.2,2.0,60.000000,3.0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...
72814,75008,NaN,NaN,2017.0,Semi-Auto,24280.0,Diesel,145.0,42.8,3.0,84.000000,0.0,0.0
74004,71229,NaN,NaN,2016.0,Manual,21762.0,Petrol,20.0,60.1,1.2,77.000000,3.0,NaN
74347,65661,NaN,NaN,2016.0,unknown,33017.0,Diesel,30.0,60.1,2.0,1.769474,4.0,0.0
74421,75855,NaN,NaN,2018.0,Semi-Auto,8980.0,Petro,NaN,42.2,1.5,74.000000,0.0,0.0


In [231]:
# Show the entries with missing values in model
missing_model_entries = x_train[x_train['model'].isnull()]
display(missing_model_entries)

,carID,Brand,model,year,transmission,mileage,fuelType,tax,mpg,engineSize,paintQuality%,previousOwners,hasDamage
11,46005,Mercedes-Benz,NaN,2020.0,Semi-Aut,1000.0,Diesel,145.000000,67.3,1.5,45.0,1.0,0.0
22,73943,Volkswagen,NaN,2016.0,Manual,29925.0,Petrol,20.000000,60.1,1.0,51.0,4.0,0.0
26,35263,Mercedes-Benz,NaN,NaN,Automatic,45000.0,Petrol,-56.193202,46.3,1.6,59.0,NaN,0.0
30,55065,Toyota,NaN,2017.0,Automatic,59000.0,Hybrid,0.000000,74.3,1.8,95.0,2.0,0.0
44,68423,Volkswagen,NaN,NaN,NaN,26941.0,Petrol,145.000000,60.1,1.2,97.0,4.0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...
75924,52995,Toyota,NaN,2019.0,automatic,6015.0,Hybrid,135.000000,57.7,1.8,NaN,4.0,0.0
75927,69479,Volkswagen,NaN,2016.0,Manual,30000.0,Petrol,20.000000,60.1,1.0,63.0,NaN,0.0
75941,69092,Volkswagen,NaN,2019.0,Automatic,32.0,Petrol,150.000000,33.2,2.0,69.0,3.0,0.0
75944,66803,Volkswagen,NaN,2017.0,Manual,50899.0,Petrol,150.000000,60.1,1.2,65.0,3.0,0.0


- Here we should decide to delete the 150 Entries with missing Brand and Model. 
- But we will keep them for now and focus on the next column 'year'.
- I will try to predict the missing model based on the other columns later, when the other columns are cleaned.

In [232]:
# Show the percentage of the entries which have a missing brand and a missing model
missing_brand_model_percentage = (x_train['Brand'].isnull() & x_train['model'].isnull()).mean() * 100
print(f"\nPercentage of entries with missing 'Brand' and missing 'model': {missing_brand_model_percentage:.2f}%")


Percentage of entries with missing 'Brand' and missing 'model': 0.20%


In [233]:
# Delete the entries missing_brand_model_percentage from x_train
# print the shape of x_train before deleting
print("\nShape of x_train before deleting entries with missing 'Brand' and 'model':", x_train.shape)
# delete the entries
x_train = x_train[~(x_train['Brand'].isnull() & x_train['model'].isnull())].reset_index(drop=True)
print("\nShape of x_train after deleting entries with missing 'Brand' and 'model':", x_train.shape)



Shape of x_train before deleting entries with missing 'Brand' and 'model': (75973, 13)

Shape of x_train after deleting entries with missing 'Brand' and 'model': (75823, 13)


In [234]:
# Look at the Missing values in x 
missing_values_x = x_train.isnull().sum()

print("\nMissing values in x_train:\n", missing_values_x[missing_values_x > 0]) 



Missing values in x_train:
 model             6715
year              1486
transmission      1519
mileage           1457
fuelType          1509
tax               7898
mpg               7923
engineSize        1515
paintQuality%     1520
previousOwners    1547
hasDamage         1544
dtype: int64


In [235]:
# Show all unique values of transmission
unique_transmissions = x_train['transmission'].unique()
print("\nUnique values in 'transmission':\n", unique_transmissions)


Unique values in 'transmission':
 ['Semi-Auto' 'Manual' 'anual' 'Semi-Aut' 'Automatic' 'manual' nan
 'unknown' 'Manua' 'AUTOMATIC' 'MANUAL' 'semi-auto' ' Manual ' 'automatic'
 'emi-Auto' 'SEMI-AUTO' 'SEMI-AUT' 'Automati' 'ANUAL' 'utomatic' 'unknow'
 'EMI-AUTO' 'manua' 'anua' 'emi-Aut' 'MANUA' 'emi-auto' 'UTOMATIC'
 ' manual ' ' MANUAL ' 'UNKNOWN' 'nknown' 'automati' 'Other' 'semi-aut'
 ' Manual' 'AUTOMATI' 'utomati' 'Manual ' 'manual ' 'nknow']


In [236]:
import pandas as pd
import numpy as np
import re

# Define a normalization function
def clean_transmission(x):
    if pd.isna(x):
        return np.nan
    s = str(x).strip().lower()

    # handle common patterns
    if re.search(r"man", s):          # e.g. 'manual', 'manua', 'manuaL '
        return "manual"
    elif re.search(r"semi", s):       # e.g. 'semi-auto', 'semi-aut'
        return "semi-auto"
    elif re.search(r"auto", s):       # e.g. 'automatic', 'automati', 'utomatic'
        return "automatic"
    elif re.search(r"unk", s):        # e.g. 'unknown', 'unknow', 'nknown'
        return "unknown"
    elif re.search(r"oth", s):        # e.g. 'other'
        return "other"
    else:
        return np.nan   # if none matches, mark as missing

# Apply it to train and test
x_train["transmission"] = x_train["transmission"].apply(clean_transmission)
x_test["transmission"] = x_test["transmission"].apply(clean_transmission)

# Optional: check results
print(x_train["transmission"].value_counts(dropna=False))
print(x_test["transmission"].value_counts(dropna=False))


transmission
manual       40735
semi-auto    16484
automatic    15193
NaN           2687
unknown        719
other            5
Name: count, dtype: int64
transmission
manual       17562
semi-auto     7148
automatic     6416
NaN           1100
unknown        338
other            3
Name: count, dtype: int64


In [237]:
# Show all unique values of fuelType
unique_fuel_types = x_train['fuelType'].unique()
print("\nUnique values in 'fuelType':\n", unique_fuel_types)


Unique values in 'fuelType':
 ['Petrol' 'Diesel' 'etrol' 'Hybrid' 'diesel' 'iesel' nan 'petrol' 'PETROL'
 'Diese' 'Petro' 'DIESEL' 'petro' 'HYBRID' 'ybri' 'Other' 'DIESE' 'Hybri'
 'ETROL' 'ybrid' 'PETRO' 'hybrid' 'IESEL' 'Electric' 'ther' 'iese' 'etro'
 'ETRO' 'diese' 'Othe' 'YBRID' 'HYBRI' 'OTHER' 'other' 'IESE']


In [238]:

def clean_fuel_type(x):
    if pd.isna(x):
        return np.nan
    s = str(x).strip().lower()

    # Match the main types
    if re.search(r"pet", s):          # petrol, petro, etrol, etc.
        return "petrol"
    elif re.search(r"die", s):        # diesel, diese, iese, etc.
        return "diesel"
    elif re.search(r"hyb", s):        # hybrid, hybr, ybrid, etc.
        return "hybrid"
    elif re.search(r"elec", s):       # electric
        return "electric"
    elif re.search(r"oth", s):        # other, othe, ther
        return "other"
    elif re.search(r"unk", s):        # if any 'unknown' appears
        return "unknown"
    else:
        return np.nan   # not recognized

# Apply cleaning to x_train and x_test
x_train["fuelType"] = x_train["fuelType"].apply(clean_fuel_type)
# Check results
print(x_train["fuelType"].value_counts(dropna=False))

x_test["fuelType"] = x_test["fuelType"].apply(clean_fuel_type)
# Check results
print(x_test["fuelType"].value_counts(dropna=False))


fuelType
petrol      40269
diesel      30200
NaN          3012
hybrid       2177
other         161
electric        4
Name: count, dtype: int64
fuelType
petrol      17151
diesel      13140
NaN          1278
hybrid        917
other          80
electric        1
Name: count, dtype: int64


In [239]:
# unique values of hasDamage
unique_has_damage = x_train['hasDamage'].unique()
print("\nUnique values in 'hasDamage':\n", unique_has_damage)


Unique values in 'hasDamage':
 [ 0. nan]


In [240]:
# count the number of True and False in hasDamage
has_damage_counts = x_train['hasDamage'].value_counts(dropna=False)
print("\nCounts of values in 'hasDamage':\n", has_damage_counts)


Counts of values in 'hasDamage':
 hasDamage
0.0    74279
NaN     1544
Name: count, dtype: int64


- We delete the Column hasDamage because it is not useful for our prediction. It only has one True value and the rest is NaN. We could 

In [241]:
# Drop the Column hasDamage from x_train and x_test
x_train = x_train.drop(columns=['hasDamage'])
x_test = x_test.drop(columns=['hasDamage'])